File: ios/Sources/Heartbeat.swift
import Foundation

#if canImport(UIKit)
    import UIKit
#endif

/// "I am still here", once an hour, whether or not there is anything to stream.
///
/// recall could not tell a dead recorder from a quiet room. Its liveness marker is
/// refreshed only by audio above the silence floor — deliberately, so a dot means
/// *recording* rather than merely connected — and while the household is paused the
/// ingest listener is closed and nothing streams at all. Capture was paused for the
/// four days before this was written, which is precisely the window in which this
/// app dying would have gone unnoticed until somebody picked the phone up (#837).
///
/// Sent to the CONTROL host (Isis, over WireGuard), not the recorder on the LAN —
/// the same split every other API call here already makes. Isis is reachable from
/// anywhere, so a phone that is out of the house still beats and "away" stops
/// looking like "dead".
///
/// ⚠ Beats only while the user has this app *started*. A stopped app is not going
/// to record, and a beat that arrived anyway would paint the one state we care
/// about — this mic will not capture anything — bright green. The check going red
/// after a deliberate Stop is correct, and pressing Start clears it.
///
/// Best-effort and silent, like `OutboxReport` on Android: a liveness report that
/// raised its own failures would be the tail wagging the dog.
enum Heartbeat {
    private static let port = 8000
    private static let timeout: TimeInterval = 8

    /// How often to beat. The grader's thresholds are expressed in multiples of this
    /// (recall.mic_alive.BEAT_EVERY_MINUTES), so the two cannot drift apart silently.
    static let every: TimeInterval = 60 * 60

    /// First retry after a beat that did not land. Doubles per consecutive failure up
    /// to `every`, so a blip costs minutes rather than an hour (#886).
    private static let retryBase: TimeInterval = 60

    /// When this process started. Read once, at first touch, which is app launch —
    /// so a beat carrying a recent value means the app restarted, and "up all week"
    /// is distinguishable from "crash-looping between beats".
    static let startedAt = Date()

    /// What one attempt did. Three cases, not two: a beat SKIPPED because the app is
    /// stopped is a deliberate state, not an unreachable control plane, and feeding it
    /// to the failure counter would spin the backoff and then beat hourly for nothing.
    enum Outcome {
        case sent
        case failed
        case skipped

        /// The failure count to carry into the next wait.
        func nextFailureCount(after current: Int) -> Int {
            switch self {
            case .sent: 0
            case .failed: current + 1
            case .skipped: current
            }
        }
    }

    /// Seconds until the next beat: the full cadence when the last one landed, a short
    /// backoff when it did not.
    ///
    /// Pure, so the schedule is pinned in tests without a network or an hour of
    /// waiting — the same reason `body` is pure.
    ///
    /// ⚠ The cap is what keeps this a BACKOFF and not a poll. One request an hour is
    /// the design; a phone that is simply off must never beat harder than that, and
    /// the whole retry burst is bounded to fit inside one cadence.
    static func nextDelay(consecutiveFailures: Int) -> TimeInterval {
        guard consecutiveFailures > 0 else { return every }
        // Doubling by multiplication rather than shifting: this counter is reset only
        // by a success, so a phone in a dead spot keeps incrementing it forever and a
        // shift would wrap.
        var delay = retryBase
        for _ in 1..<max(consecutiveFailures, 1) {
            if delay >= every { return every }
            delay *= 2
        }
        return min(delay, every)
    }

    /// App version and build, so a restart *into a new build* reads as a deploy
    /// rather than as a fault.
    static var version: String {
        let info = Bundle.main.infoDictionary
        let short = info?["CFBundleShortVersionString"] as? String ?? "?"
        let build = info?["CFBundleVersion"] as? String ?? "?"
        return "\(short) (\(build))"
    }

    /// The JSON a beat carries. Pure and separate from the send so it can be tested
    /// without a network — the field names are a contract with `HeartbeatIn`.
    static func body(
        device: String, version: String, startedAt: Date, streaming: Bool, charging: Bool?,
        micOk: Bool
    ) -> [String: Any] {
        var out: [String: Any] = [
            "device": device,
            "app": "ios",
            "version": version,
            "startedAt": iso(startedAt),
            "streaming": streaming,
            // A running app that cannot open its mic must SAY so rather than fall
            // silent, which is what it used to do (#887).
            "micOk": micOk,
        ]
        // Absent rather than null when unknown: the simulator and a device with
        // battery monitoring off both report `.unknown`, and guessing "discharging"
        // there would invent the very reading a room phone is watched for.
        if let charging { out["charging"] = charging }
        return out
    }

    /// True on mains, false on battery, nil when iOS will not say. Room phones are
    /// mains-powered, so discharging is the leading indicator of the death this
    /// whole feature exists to catch — carried, though never graded, because a
    /// carried phone is off charge all day and that is not a fault.
    static func charging() -> Bool? {
        #if canImport(UIKit)
            UIDevice.current.isBatteryMonitoringEnabled = true
            switch UIDevice.current.batteryState {
            case .charging, .full: return true
            case .unplugged: return false
            default: return nil
            }
        #else
            return nil
        #endif
    }

    /// POST one beat, trying the control plane first and the recorder's LAN address
    /// second (#888).
    ///
    /// The fallback exists because the beat used to demand MORE reachability than
    /// recording does: audio goes to `host` on the LAN, so a phone at home with its
    /// tunnel off records every sample and still read as dead. `lanHost` runs the
    /// relay on the same port, so this is the identical request with the host
    /// swapped — and the relay marks what it forwards, so "alive but its tunnel is
    /// down" stays visible instead of being papered over.
    ///
    /// Order matters: the VPN is tried FIRST so the LAN path is a backstop rather
    /// than a shortcut, and a phone away from home behaves exactly as before.
    @discardableResult
    static func send(
        host: String, lanHost: String = "", device: String, streaming: Bool, micOk: Bool
    ) async -> Bool {
        let payload = body(
            device: device, version: version, startedAt: startedAt,
            streaming: streaming, charging: charging(), micOk: micOk)
        for candidate in [host, lanHost] where !candidate.isEmpty {
            if await post(payload, to: candidate) { return true }
        }
        return false
    }

    private static func post(_ payload: [String: Any], to host: String) async -> Bool {
        guard let url = URL(string: "http://\(host):\(port)/api/devices/heartbeat")
        else { return false }
        guard let data = try? JSONSerialization.data(withJSONObject: payload) else {
            return false
        }
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.timeoutInterval = timeout
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = data
        guard let (_, resp) = try? await URLSession.shared.data(for: req) else {
            return false
        }
        // Any 2xx: the fleet answers 200, the LAN relay 204 (it stores nothing of
        // its own, so it has no body to return). Insisting on 200 would have made
        // every relayed beat read as a failure.
        let code = (resp as? HTTPURLResponse)?.statusCode ?? 0
        return (200..<300).contains(code)
    }

    private static func iso(_ date: Date) -> String {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f.string(from: date)
    }
}

File: android/app/src/main/kotlin/org/recall/mic/Heartbeat.kt
package org.recall.mic

import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.BatteryManager
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.time.Instant
import java.time.format.DateTimeFormatter

/**
 * "I am still here", once an hour, whether or not there is anything to stream.
 *
 * recall could not tell a dead recorder from a quiet room. Its per-source liveness marker
 * is refreshed only by audio above the silence floor — deliberately, so that a dot means
 * *recording* rather than merely connected — and while the household is paused the ingest
 * listener is closed and nothing streams at all. Capture was paused for the four days
 * before this was written, which is exactly the window in which a dead app goes unnoticed
 * until somebody picks the phone up (#837).
 *
 * Sent to the CONTROL host (Isis, over WireGuard), not the recorder on the LAN — the same
 * split [CaptureApi] already makes. Isis is reachable from anywhere, so a phone that is
 * out of the house still beats and "away" stops looking like "dead". That is also why
 * this is not merged into [OutboxReport]: that one rides the upload worker's schedule and
 * only exists on phones that record meetings; this must beat from the streaming service,
 * on every mic phone, on its own clock.
 *
 * ⚠ Beats only while the service is running, which is what "started" means here. A
 * stopped app is not going to record, and a beat that arrived anyway would paint the one
 * state worth catching — this mic will capture nothing — bright green.
 *
 * Best-effort and silent, like [OutboxReport]: a liveness report that raised its own
 * failures would be the tail wagging the dog.
 */
object Heartbeat {
    private const val API_PORT = 8000 // `recall api --port 8000`, as CaptureApi
    private const val TIMEOUT_MS = 8000

    /**
     * How often to beat. The grader's thresholds are written as multiples of this
     * (`recall.mic_alive.BEAT_EVERY_MINUTES`), so the two cannot drift apart silently.
     */
    const val EVERY_MINUTES = 60L

    /**
     * First retry after a beat that did not land. Doubles per consecutive failure up to
     * [EVERY_MINUTES], so a blip costs minutes rather than an hour (#886).
     */
    private const val RETRY_BASE_MINUTES = 1L

    /**
     * When this process started. Set on class load — the service is the first thing to
     * touch it — so a beat carrying a recent value means the app restarted, and "up all
     * week" is distinguishable from "relaunching between beats".
     */
    val startedAt: Instant = Instant.now()

    /**
     * Minutes until the next beat: the full cadence when the last one landed, a short
     * backoff when it did not.
     *
     * Pure, so the schedule is pinned in unit tests without a network or an hour of
     * waiting — the same reason [body] is pure.
     *
     * ⚠ The cap is what keeps this a BACKOFF and not a poll. One request an hour is the
     * design; a phone that is simply off (or out of range for a week) must never beat
     * harder than that, and the whole retry burst is bounded to fit inside one cadence.
     * Growing it would trade the thing this exists to protect for a little less latency.
     */
    fun nextDelayMinutes(consecutiveFailures: Int): Long {
        if (consecutiveFailures <= 0) return EVERY_MINUTES
        // Shift-free: `1L shl 63` would wrap, and this counter is reset only by a
        // success, so a phone left in a dead spot keeps incrementing it forever.
        var delay = RETRY_BASE_MINUTES
        repeat(consecutiveFailures - 1) {
            if (delay >= EVERY_MINUTES) return EVERY_MINUTES
            delay *= 2
        }
        return minOf(delay, EVERY_MINUTES)
    }

    /** App version and build, so a restart *into a new build* reads as a deploy. */
    fun version(ctx: Context): String =
        runCatching {
            val info = ctx.packageManager.getPackageInfo(ctx.packageName, 0)
            @Suppress("DEPRECATION")
            "${info.versionName} (${info.versionCode})"
        }.getOrDefault("?")

    /**
     * The JSON a beat carries — pure, so the field names (a contract with the server's
     * `HeartbeatIn`) can be pinned in a unit test without a network.
     */
    fun body(
        device: String,
        version: String,
        startedAt: Instant,
        streaming: Boolean,
        charging: Boolean?,
        micOk: Boolean,
    ): String =
        JSONObject()
            .put("device", device)
            .put("app", "android")
            .put("version", version)
            .put("startedAt", DateTimeFormatter.ISO_INSTANT.format(startedAt))
            .put("streaming", streaming)
            // A running app that cannot open its mic must SAY so rather than fall
            // silent, which is what it used to do (#887).
            .put("micOk", micOk)
            // Absent rather than null when unknown: guessing "discharging" would invent
            // the very reading a mains-powered room phone is watched for.
            .apply { if (charging != null) put("charging", charging) }
            .toString()

    /**
     * True on mains, false on battery, null when Android will not say. Room phones are
     * mains-powered, so discharging is the leading indicator of the death this exists to
     * catch — carried, never graded, because a carried phone is off charge all day.
     */
    fun charging(ctx: Context): Boolean? =
        runCatching {
            val status =
                ctx
                    .registerReceiver(null, IntentFilter(Intent.ACTION_BATTERY_CHANGED))
                    ?.getIntExtra(BatteryManager.EXTRA_STATUS, -1)
                    ?: -1
            when (status) {
                BatteryManager.BATTERY_STATUS_CHARGING,
                BatteryManager.BATTERY_STATUS_FULL,
                -> true

                BatteryManager.BATTERY_STATUS_DISCHARGING,
                BatteryManager.BATTERY_STATUS_NOT_CHARGING,
                -> false

                else -> null
            }
        }.getOrNull()

    /**
     * POST one beat, blocking. Returns whether it landed; nothing depends on it.
     *
     * Blocking rather than `suspend` on purpose: the caller is [StreamService]'s own beat
     * thread, which has no coroutine scope and whose whole job is to sleep and send.
     */
    fun send(
        controlHost: String,
        device: String,
        streaming: Boolean,
        micOk: Boolean,
        ctx: Context,
    ): Boolean =
        runCatching {
            if (controlHost.isBlank()) return false
            val body = body(device, version(ctx), startedAt, streaming, charging(ctx), micOk)
            val conn =
                (
                    URL("http://$controlHost:$API_PORT/api/devices/heartbeat")
                        .openConnection() as HttpURLConnection
                ).apply {
                    requestMethod = "POST"
                    doOutput = true
                    connectTimeout = TIMEOUT_MS
                    readTimeout = TIMEOUT_MS
                    setRequestProperty("Content-Type", "application/json")
                }
            conn.outputStream.use { it.write(body.toByteArray()) }
            val code = conn.responseCode
            conn.disconnect()
            code in 200..299
        }.getOrDefault(false)
}
